storage/src/cache/dedup/db.rs

```diff
@@ -0,0 +1,329 @@
+        };
+
+        Ok(None)
+    }
+
+    pub fn get_all_blobs(&self) -> Result<Vec<(u64, String)>> {
+        let conn = self.get_connection()?;
+        let mut stmt = conn.prepare_cached("SELECT BlobId, FilePath FROM Blobs")?;
+        let rows = stmt.query_map([], |row| Ok((row.get::<usize, u64>(0)?, row.get(1)?)))?;
+        let mut results: Vec<(u64, String)> = Vec::new();
+        for row in rows {
+            results.push(row?);
+        }
+        Ok(results)
+    }
+
+    pub fn add_blobs(&mut self, blobs: &[String]) -> Result<()> {
+        let sql = "INSERT OR IGNORE INTO Blobs (FilePath) VALUES (?1)";
+        let mut conn = self.get_connection()?;
+        let tran = Self::begin_transaction(&mut conn)?;
+
+        for blob in blobs {
+            if let Err(e) = tran.execute(sql, [blob]) {
+                return Err(e.into());
+            };
+        }
+        tran.commit()?;
+
+        Ok(())
+    }
+
+    pub fn add_blob(&self, blob: &str) -> Result<u64> {
+        let sql = "INSERT OR IGNORE INTO Blobs (FilePath) VALUES (?1)";
+        let conn = self.get_connection()?;
+        conn.execute(sql, [blob])?;
+        Ok(conn.last_insert_rowid() as u64)
+    }
+
+    pub fn delete_blobs(&mut self, blobs: &[String]) -> Result<()> {
+        let delete_blobs_sql = "DELETE FROM Blobs WHERE FilePath = (?1)";
+        let delete_chunks_sql = "DELETE FROM Chunks WHERE BlobId = (?1)";
+        let mut conn = self.get_connection()?;
+        let tran = Self::begin_transaction(&mut conn)?;
+
+        for blob in blobs {
+            if let Some(id) = Self::get_blob_id_with_tx(&tran, blob)? {
+                if let Err(e) = tran.execute(delete_chunks_sql, [id]) {
+                    return Err(e.into());
+                }
+            }
+            if let Err(e) = tran.execute(delete_blobs_sql, [blob]) {
+                return Err(e.into());
+            }
+        }
+        tran.commit()?;
+
+        Ok(())
+    }
+
+    pub fn get_chunk_info(&self, chunk_id: &str) -> Result<Option<(String, u64)>> {
+        let sql = "SELECT FilePath, ChunkOffset \
+                FROM Chunks INDEXED BY ChunkIndex \
+                JOIN Blobs ON Chunks.BlobId = Blobs.BlobId \
+                WHERE ChunkId = ?\
+                ORDER BY Blobs.BlobId LIMIT 1 OFFSET 0";
+
+        if let Some((new_blob_id, chunk_info)) = self
+            .get_connection()?
+            .query_row(sql, [chunk_id], |row| {
+                Ok((row.get(0)?, row.get::<usize, u64>(1)?))
+            })
+            .optional()?
+        {
+            return Ok(Some((new_blob_id, chunk_info)));
+        }
+
+        Ok(None)
+    }
+
+    pub fn add_chunks(&mut self, chunks: &[(String, u64, String)]) -> Result<()> {
+        let sql = "INSERT OR IGNORE INTO Chunks (ChunkId, ChunkOffset, BlobId) VALUES (?1, ?2, ?3)";
+        let mut conn = self.get_connection()?;
+        let tran = Self::begin_transaction(&mut conn)?;
+
+        for chunk in chunks {
```

Comment:
Can we use batch insert here?

Reply:
How about https://stackoverflow.com/questions/1609637/how-to-insert-multiple-rows-in-sqlite